src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+        WideInt b = 43;
+        a = b;
+        test!("==")(a, 43);
+    }
+
+    /***************************************************************************
+
+        Enables equality comparison with a plain integer type
+
+        Params:
+            rhs = value to compare to
+
+    ***************************************************************************/
+
+    public equals_t opEquals ( ulong rhs )
+    {
+        if (this.payload[0] != (rhs & uint.max))
+            return false;
+
+        if (this.payload[1] != (rhs >> 32))
+            return false;
+
+        foreach (ref word; this.payload[2 .. N])
+        {
+            if (word != 0)
+                return false;
+        }
+
+        return true;
+    }
+
+    unittest
+    {
+        WideInt i;
+
+        i = 42;
+        test!("==")(i, 42);
+
+        i = ulong.max / 2;
+        test!("==")(i, ulong.max / 2);
+        test!("!=")(i, ulong.max);
+
+        i.payload[$-1] = 42;
+        test!("!=")(i, 42);
+    }
+
+    /***************************************************************************
+
+        Enables equality comparison with WideInt of the same size
+
+        Params:
+            rhs = value to compare against
+
+    ***************************************************************************/
+
+    public equals_t opEquals ( WideInt rhs )
+    {
+        return this.payload[] == rhs.payload[];
+    }
+
+    unittest
+    {
+        WideInt a, b;
+        test!("==")(a, b);
+        foreach (size_t i, ref elem; a.payload)
+        {
+            a.payload[i] = 1337;
+            b.payload[i] = 1337;
+        }
+        test!("==")(a, b);
+        a.payload[$-1] = 0;
+        test!("!=")(a, b);
+    }
+
+    /***************************************************************************
+
+        Increment current value by one fitting in uint
+
+        Params:
+            rhs = value to increment with. Limited to `uint` for now to simplify
+                internal arithmetic.
+
+        Throws:
+            WideIntRangeException if this WideInt value range was to be
+            exceeded
+
+    ***************************************************************************/
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
+            enforce(.wideint_exception, checkAndInc(1));
+    }
+
+    unittest
+    {
+        WideInt num = 0;
+        for (auto i = 0; i < 1000; ++i)
+            num += (1 << 31);
+        test!("==")(num.payload[0], 0);
+        test!("==")(num.payload[1], 500);
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
```

Comment:
No, this is simply a most basic functional primitive to build other API on top

As for `WideInt rhs` overloads - do we actually need them now? I assumed it will be used as sort of aggregator value which only works with regular `ulong`/`uint` values individually but may internally grow over their size limit.

My intention is to implement minimal amount of functionality that fits current purpose and leave rest to PRs of interested parties ;)